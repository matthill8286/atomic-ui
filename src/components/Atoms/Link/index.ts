export * from './Link'
export * from './Link.interface'
