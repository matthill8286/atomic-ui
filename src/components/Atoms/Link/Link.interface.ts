import * as H from 'history'
import React, { ComponentType, MouseEvent, MouseEventHandler, ReactElement, ReactNode } from 'react'
import { ThemeColors, ThemeFontLineHeight, ThemeFontSizes } from '@/types/theme'
import { ButtonType } from '@/components/Atoms/Button'

export type LinkTarget = string | undefined

export type LinkHref = string | undefined

export interface WrappedRouterLinkProps {
  children: ReactNode
  className?: string
  onClick?: (e: MouseEvent<HTMLElement>) => void
  onMouseDown?: (e: MouseEvent) => void
  to?: string
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type LinkAs = string | JSX.IntrinsicElements | ComponentType<any>
export type LinkSize = 'small' | 'large'
type CursorType = 'initial' | 'pointer'

export interface LinkProps
  extends LinkRouterProps<HTMLAnchorElement | HTMLButtonElement>,
    React.AnchorHTMLAttributes<HTMLAnchorElement | HTMLButtonElement> {
  bold?: boolean
  cursor?: CursorType
  color?: ThemeColors
  decorationColor?: ThemeColors
  fontSize?: ThemeFontSizes
  iconLeft?: ReactElement
  iconRight?: ReactElement
  inline?: boolean
  lineHeight?: ThemeFontLineHeight
  scale?: LinkSize
  scrollOffset?: number
  smooth?: boolean
  underline?: boolean
  disabled?: boolean
  type?: ButtonType
  target?: LinkTarget
  href?: LinkHref
  branded?: boolean
  ariaLabel?: string
  isGrouped?: boolean
}

export interface LinkRouterProps<T> {
  to?: H.LocationDescriptor
  children?: React.ReactNode
  className?: string
  onClick?: MouseEventHandler<T>
}

export interface StyledLinkProps {
  color?: ThemeColors
  cursor?: CursorType
  hasIconLeft: boolean
  hasIconRight: boolean
  isBold: boolean
  isGrouped?: boolean
  isDisabled: boolean
  isInline: boolean
  isLarge: boolean
  isSmall: boolean
  isUnderlined?: boolean
  target?: string
  title?: string
}
