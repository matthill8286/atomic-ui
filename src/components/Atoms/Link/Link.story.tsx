import { boolean, number, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { BrowserRouter } from 'react-router-dom'
import { Icon } from '@/components/Atoms/Icon'
import { CopyText } from '@/components/Atoms/Typography'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { OtherShare, OtherVideo } from '@matthill8286/atomic-icon-library'
import { ThemeColors, ThemeFontSizes } from '@/types/theme'
import { Link } from './Link'
import { LinkProps, LinkSize } from './Link.interface'
import { OtherEdit } from '@matthill8286/atomic-icon-library'
import { action } from '@storybook/addon-actions'

const scales = ['large', 'small']
const sizes: ThemeFontSizes[] = ['xxxs', 'xxs', 'xs', 'sm', 'md', 'lg', 'xl', 'xxl', 'xxxl']
const targets: string[] = ['_self', '_blank']

const knobs = (): Partial<LinkProps> => {
  return {
    color: select('Color', Object.keys(saiyanTheme.color), 'grey4') as ThemeColors,
    disabled: boolean('Set disabled', false),
    bold: boolean('Bold', false),
    target: select('Target', targets, '_self'),
  }
}

storiesOf('Design System/Atoms/Link', module)
  .add('Link', () => {
    const scale = select('Type scale', scales, 'large') as LinkSize
    const sampleText = text('Example text', `Lorem ipsum dolor`)
    const underline = boolean('Underline', true)

    return (
      <Link {...knobs()} scale={scale} href="#" underline={underline}>
        {sampleText}
      </Link>
    )
  })
  .add('with Icon', () => {
    const scale = select('Type scale', scales, 'large') as LinkSize
    const sampleText = text('Example text', `Lorem ipsum dolor`)
    const iconPositions = ['left', 'right', 'both']
    const iconPosition = select('Icon Position', iconPositions, 'left')
    const iconColor = select('Icon Color', Object.keys(saiyanTheme.color), 'primary') as ThemeColors
    const underline = boolean('Underline', true)

    return (
      <BrowserRouter>
        <Link
          {...knobs()}
          to="/test"
          scale={scale}
          underline={underline}
          {...((iconPosition === 'left' || iconPosition === 'both') && {
            iconLeft: (
              <Icon color={iconColor}>
                <OtherShare height="sm" width="sm" fill={iconColor} />
              </Icon>
            ),
          })}
          {...((iconPosition === 'right' || iconPosition === 'both') && {
            iconRight: (
              <Icon color={iconColor}>
                <OtherEdit height="sm" width="sm" fill={iconColor} />
              </Icon>
            ),
          })}>
          {sampleText}
        </Link>
      </BrowserRouter>
    )
  })
  .add('inline Link', () => {
    const sampleText = text('Example text', `Lorem ipsum link`)
    const fontSize = select('Size', sizes, sizes[3])
    const underline = boolean('Underline', true)
    return (
      <CopyText fontSize={fontSize} tag="div">
        <Link {...knobs()} fontSize={fontSize} inline href="/test" underline={underline}>
          {sampleText}
        </Link>
        {' found this useful'}
      </CopyText>
    )
  })
  .add('smooth scroll Anker Link', () => {
    const sampleText = text('Example text', `Lorem ipsum link`)
    const fontSize = select('Size', sizes, sizes[3])
    const scrollOffset = number('ScrollOffset', 0)
    return (
      <div>
        <CopyText fontSize={fontSize} tag="div">
          <Link
            {...knobs()}
            fontSize={fontSize}
            href="#smooth_anker"
            smooth
            scrollOffset={scrollOffset}>
            {sampleText}
          </Link>
        </CopyText>
        <div id="smooth_anker" style={{ marginTop: '1500px', marginBottom: '800px' }}>
          Anker Target
        </div>
      </div>
    )
  })
  .add('Link as span', () => {
    const scale = select('Type scale', scales, 'large') as LinkSize
    const sampleText = text('Example text', `Lorem ipsum dolor`)
    const underline = boolean('Underline', true)

    return (
      <Link {...knobs()} scale={scale} href="/test" underline={underline}>
        {sampleText}
      </Link>
    )
  })
  .add('Link as branded', () => {
    const scale = select('Type scale', scales, 'large') as LinkSize
    const sampleText = text('Example text', `Lorem ipsum dolor`)
    const branded = boolean('Branded', true)

    return (
      <Link {...knobs()} underline={false} scale={scale} href="#" branded={branded}>
        {sampleText}
      </Link>
    )
  })

const buttonKnobs = (): Partial<LinkProps> => {
  return {
    color: select('Color', Object.keys(saiyanTheme.color), 'black') as ThemeColors,
    disabled: boolean('Set disabled', false),
    bold: boolean('Bold', false),
    onClick: action('onclick'),
  }
}

storiesOf('Design System/Atoms/Link/Button', module)
  .add('Default', () => {
    const scale = select('Type scale', scales, 'large') as LinkSize
    const sampleText = text('Example text', `Lorem ipsum dolor`)
    const underline = boolean('Underline', true)

    return (
      <Link {...buttonKnobs()} scale={scale} underline={underline}>
        {sampleText}
      </Link>
    )
  })
  .add('With Icon', () => {
    const scale = select('Type scale', scales, 'large') as LinkSize
    const sampleText = text('Example text', `Lorem ipsum dolor`)
    const iconPositions = ['left', 'right', 'both']
    const iconPosition = select('Icon Position', iconPositions, 'left')
    const iconColor = select('Icon Color', Object.keys(saiyanTheme.color), 'primary') as ThemeColors
    const underline = boolean('Underline', true)

    return (
      <Link
        {...buttonKnobs()}
        scale={scale}
        underline={underline}
        {...((iconPosition === 'left' || iconPosition === 'both') && {
          iconLeft: (
            <Icon color={iconColor}>
              <OtherShare height="sm" width="sm" fill={iconColor} />
            </Icon>
          ),
        })}
        {...((iconPosition === 'right' || iconPosition === 'both') && {
          iconRight: (
            <Icon color={iconColor}>
              <OtherEdit height="sm" width="sm" fill={iconColor} />
            </Icon>
          ),
        })}>
        {sampleText}
      </Link>
    )
  })
  .add('Inline', () => {
    const sampleText = text('Example text', `Lorem ipsum link`)
    const fontSize = select('Size', sizes, sizes[3])
    const underline = boolean('Underline', true)

    return (
      <CopyText fontSize={fontSize} tag="div">
        {'This is random Text with a '}
        <Link {...buttonKnobs()} fontSize={fontSize} inline underline={underline}>
          {sampleText}
        </Link>
        {' read more...'}
      </CopyText>
    )
  })
