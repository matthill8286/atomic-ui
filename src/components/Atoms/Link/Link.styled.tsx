import React, { FunctionComponent } from 'react'
import { Link as RouterLink } from 'react-router-dom'
import { CopyText } from '@/components/Atoms/Typography'
import { css, styled } from '@/styles/styled'
import { Theme, ThemeColors } from '@/types/theme'
import { LinkRouterProps, StyledLinkProps } from './Link.interface'

export const linkStyles = css<StyledLinkProps>`
  display: ${({ isInline }) => (isInline ? `inline-flex` : 'flex')};
  align-items: center;
  width: max-content;
  color: ${({ theme, isDisabled, isInline, color }) =>
    (isDisabled && theme.color.grey2) || (isInline && 'inherit') || (color && theme.color[color])};
  opacity: ${({ theme, isDisabled }) => (isDisabled ? theme.polished.inactive : `1.0`)};
  text-decoration: none;
  cursor: ${({ isDisabled, cursor }) => (isDisabled ? 'initial' : cursor || 'pointer')};
  position: relative;
  border-bottom: ${({ isUnderlined, isInline }) => (isUnderlined && isInline ? '1px solid' : '')};
  ${({ theme, color, isInline }) =>
    isInline && color && `border-bottom-color: ${theme.color[color]}`};
  padding: ${({ theme, isInline, isGrouped }) =>
    !isInline || isGrouped ? theme.spacing.base.sm : 0};

  &::after {
    content: '';
    position: absolute;
    display: ${({ isUnderlined }) => (isUnderlined ? 'block' : 'none')};
    bottom: -2px;
    height: ${({ isBold }) => (isBold ? 2 : 1)}px;
    ${({ isUnderlined }) => (isUnderlined ? widthAndPosForUnderline : '')}
    background-color: ${({ theme, color, isDisabled }) =>
      (isDisabled && theme.color.grey2) || (color && theme.color[color]) || theme.color.primary};
    transform-origin: right center;
    transform: scale(0, 1);
    transition: transform 0.25s cubic-bezier(0.37, 0.31, 0.2, 0.85);
  }
  ${({ isDisabled }) =>
    !isDisabled &&
    `
    :hover {
      &::after {
        transform-origin: left center;
        transform: scale(1, 1);
      }
    }
  `}
`

export const StyledLinkAnchor = styled.a<StyledLinkProps>`
  ${linkStyles};
`

export const StyledLinkButton = styled.button<StyledLinkProps>`
  appearance: none;
  background: transparent;
  border: 0;
  padding: 0;

  ${linkStyles};

  /* Normalize line-height. Cannot be changed from normal in Firefox 4+. */
  line-height: normal;

  /* Corrects font smoothing for webkit */
  -webkit-font-smoothing: inherit;
  -moz-osx-font-smoothing: inherit;

  /* Corrects inability to style clickable input types in iOS */
  -webkit-appearance: none;

  /* Remove excess padding and border in Firefox 4+ */
  &::-moz-focus-inner {
    border: 0;
    padding: 0;
  }
`

const RouterLinkWrapper: FunctionComponent<LinkRouterProps<
  HTMLAnchorElement | HTMLButtonElement
>> = ({ children, className, to = '#', onClick }) => (
  <RouterLink onClick={onClick} className={className} to={to}>
    {children}
  </RouterLink>
)

export const StyledLinkRouter = styled(RouterLinkWrapper)`
  ${linkStyles};
`

export const StyledIconWrapper = styled.span<StyledLinkProps>`
  margin-left: ${({ hasIconRight, theme }) => (hasIconRight ? getIconMargin(theme) : '0px')};
  margin-right: ${({ hasIconLeft, theme }) => (hasIconLeft ? getIconMargin(theme) : '0px')};

  & svg {
    height: ${({ theme, isSmall }) => getIconSize(theme, isSmall)};
    width: ${({ theme, isSmall }) => getIconSize(theme, isSmall)};
  }
`

export const Text = styled(CopyText)<{
  isInline: boolean
  isBranded?: boolean
  color?: ThemeColors
}>`
  display: ${props => (props.isInline ? 'inline' : 'block')};
  background-color: ${({ theme, isBranded }) => isBranded && theme.color.primary};
  color: ${({ theme, color }) => (color && theme.color[color]) || theme.color.grey2};
  padding: ${({ theme, isBranded }) =>
    isBranded && `${theme.spacing.base.xs} ${theme.spacing.base.sm}`};
  border-radius: ${({ isBranded, theme }) => isBranded && theme.dimension.borderRadius6};
`

/**
 * This helper function calculates width of the underline.
 * The width is not 100% if the link has an icon. If the link has an icon we also need a shift to left or right.
 * If no icon is give, just return the normal case which the underline has the full width.
 */
const widthAndPosForUnderline = css<StyledLinkProps>(
  ({ theme, isSmall, hasIconLeft, hasIconRight }) => {
    if (!hasIconLeft && !hasIconRight) {
      return css`
        left: 0;
        width: calc(100%);
      `
    }

    const leftOffset = getIconWidth(theme, isSmall)
    const iconWidth = getIconWidth(theme, isSmall, hasIconLeft && hasIconRight)

    const iconLeft = css`
      left: ${leftOffset};
    `
    return css`
      ${hasIconLeft && iconLeft}

      width: calc(
            100% - ${iconWidth}
          );
    `
  }
)

/**
 * Get the width/height of the icon depending if the link is small or large.
 *
 * @param theme
 * @param isSmall
 */
const getIconSize = (theme: Theme, isSmall: boolean) => {
  return isSmall ? theme.font.lineHeight.xxs : theme.font.lineHeight.xs
}

/**
 * Returns the IconWidth including the margin.
 * If calculateForTwoIcons is true, it double the icon width, since there are two icons (left and right)
 * The margin is between icon and text.
 *
 * @param theme
 * @param isSmall
 * @param calculateForTwoIcons
 */
const getIconWidth = (theme: Theme, isSmall: boolean, calculateForTwoIcons?: boolean) => {
  const iconSizeAsNum = parseInt(getIconSize(theme, isSmall))
  const iconMarginAsNum = parseInt(getIconMargin(theme))
  return (iconSizeAsNum + iconMarginAsNum) * (calculateForTwoIcons ? 2 : 1) + 'px'
}

const getIconMargin = (theme: Theme) => {
  return theme.spacing.base.xxs
}
