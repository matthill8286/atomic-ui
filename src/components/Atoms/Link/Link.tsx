import React from 'react'
import { Link as RouterLink } from 'react-router-dom'
import { smoothScrollWithEvent } from '@/utils'
import { LinkProps, StyledLinkProps } from './Link.interface'
import {
  StyledIconWrapper,
  StyledLinkAnchor,
  StyledLinkButton,
  StyledLinkRouter,
  Text,
} from './Link.styled'

export const Link: React.FC<LinkProps> = ({
  bold = false,
  branded = false,
  children,
  className,
  cursor,
  color = 'grey4',
  decorationColor,
  disabled = false,
  fontSize,
  href,
  iconLeft,
  iconRight,
  inline = false,
  lineHeight,
  onClick,
  onMouseDown,
  scale = 'small',
  scrollOffset = 0,
  smooth = false,
  target = '_self',
  title,
  to,
  underline = true,
  isGrouped = false,
  ariaLabel,
  ...otherProps
}): JSX.Element => {
  const isSmall = scale === 'small'
  const isButton = onClick && !href && !to
  const isRouterLink = to

  const styleProps: StyledLinkProps = {
    color: decorationColor || color,
    cursor,
    hasIconLeft: !!iconLeft,
    hasIconRight: !!iconRight,
    isBold: bold,
    isDisabled: disabled,
    isInline: inline,
    isLarge: scale === 'large',
    isSmall,
    isUnderlined: underline,
    isGrouped,
  }
  let onClickWrapper

  if (smooth) {
    onClickWrapper = (event: React.MouseEvent<HTMLAnchorElement | HTMLButtonElement>) => {
      try {
        smoothScrollWithEvent(event, scrollOffset)
      } catch (e) {
        //failsafe
      }
      if ('function' == typeof onClick) {
        return onClick(event)
      }
    }
  } else {
    onClickWrapper = onClick
  }

  const getChildren = (): JSX.Element => (
    <>
      {iconLeft && <StyledIconWrapper {...styleProps}>{iconLeft}</StyledIconWrapper>}
      {typeof children === 'string' ? (
        <Text
          tag="span"
          fontSize={fontSize || (isSmall ? 'xs' : 'sm')}
          lineHeight={lineHeight}
          bold={bold}
          color={color}
          isBranded={branded}
          isInline={inline}
          aria-label={ariaLabel}>
          {children}
        </Text>
      ) : (
        children
      )}
      {iconRight && <StyledIconWrapper {...styleProps}>{iconRight}</StyledIconWrapper>}
    </>
  )

  const linkType = (): JSX.Element => {
    if (isButton) {
      return (
        <StyledLinkButton
          className={className}
          {...(disabled ? {} : { onClick: onClickWrapper })}
          {...styleProps}
          {...otherProps}>
          {getChildren()}
        </StyledLinkButton>
      )
    }

    if (isRouterLink) {
      return (
        <StyledLinkRouter
          onClick={onClickWrapper}
          to={to}
          className={className}
          {...styleProps}
          data-test="ewb-router-link">
          {getChildren()}
        </StyledLinkRouter>
      )
    }

    return (
      <StyledLinkAnchor
        className={className}
        target={target}
        rel={target === '_blank' ? 'noopener noreferrer' : otherProps.rel}
        {...(disabled ? {} : { href, onClick: onClickWrapper })}
        {...styleProps}
        {...otherProps}>
        {getChildren()}
      </StyledLinkAnchor>
    )
  }

  return linkType()
}
