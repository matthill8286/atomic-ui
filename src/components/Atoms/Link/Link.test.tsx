import React from 'react'
import { Icon } from '@/components/Atoms/Icon'
import { OtherArrow } from '@matthill8286/atomic-icon-library'
import {
  mountWithTheme,
  mountWithThemeAndRouter,
  renderWithTheme,
  renderWithThemeAndRouter,
} from '@/testRenderer'
import { Link } from './Link'

describe('<Link /> component', () => {
  const onClick = jest.fn()
  const linkContent = 'Button'
  const linkHref = 'https://google.com/?how-to-write-tests'

  it('should render a button tag against the "onClick" attribute', () => {
    const button = renderWithTheme(<Link onClick={onClick}>{linkContent}</Link>)
    const tree = mountWithTheme(<Link onClick={onClick}>{linkContent}</Link>)
    expect(tree.find('button')).toHaveLength(1)
    expect(button).toMatchSnapshot()
  })

  it('should not render a button tag against the "href" attribute', () => {
    const tree = mountWithTheme(<Link href={linkHref}>{linkContent}</Link>)
    expect(tree.find('button')).toHaveLength(0)
  })

  it('should not render a button tag against the "to" attribute', () => {
    const tree = mountWithThemeAndRouter(<Link to="/">{linkContent}</Link>)
    expect(tree.find('button')).toHaveLength(0)
  })

  it('should render an anchor tag against the "href" attribute', () => {
    const anchor = renderWithTheme(<Link href={linkHref}>{linkContent}</Link>)
    const tree = mountWithTheme(<Link href={linkHref}>{linkContent}</Link>)
    expect(tree.find('a')).toHaveLength(1)
    expect(anchor).toMatchSnapshot()
  })

  it('should not render an anchor tag against the "onClick" attribute', () => {
    const tree = mountWithTheme(<Link onClick={onClick}>{linkContent}</Link>)
    expect(tree.find('a')).toHaveLength(0)
  })

  it('should render an anchor tag from a router link against the "to" attribute', () => {
    const routerLink = renderWithThemeAndRouter(<Link to="/">{linkContent}</Link>)
    const tree = mountWithThemeAndRouter(<Link to="/">{linkContent}</Link>)
    expect(tree.find('RouterLinkWrapper[data-test="ewb-router-link"]')).toHaveLength(1)
    expect(routerLink).toMatchSnapshot()
  })

  it('should not render an anchor tag from a router link against the "onClick" attribute', () => {
    const tree = mountWithTheme(<Link onClick={onClick}>{linkContent}</Link>)
    expect(tree.find('RouterLinkWrapper[data-test="ewb-router-link"]')).toHaveLength(0)
  })

  it('should not render an anchor tag from a router link against the "href" attribute', () => {
    const tree = mountWithTheme(<Link href={linkHref}>{linkContent}</Link>)
    expect(tree.find('RouterLinkWrapper[data-test="ewb-router-link"]')).toHaveLength(0)
  })

  it('should render an anchor tag with a left icon', () => {
    const tree = renderWithTheme(
      <Link
        scale="large"
        iconLeft={
          <Icon>
            <OtherArrow />
          </Icon>
        }>
        {linkContent}
      </Link>
    )
    expect(tree).toMatchSnapshot()
  })

  it('should render an anchor tag with a right icon', () => {
    const tree = renderWithTheme(
      <Link
        scale="small"
        iconRight={
          <Icon>
            <OtherArrow />
          </Icon>
        }>
        {linkContent}
      </Link>
    )
    expect(tree).toMatchSnapshot()
  })
})
