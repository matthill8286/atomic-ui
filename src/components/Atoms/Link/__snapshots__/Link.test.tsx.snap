// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Link /> component should render a button tag against the "onClick" attribute 1`] = `
.c1 {
  color: #999999;
  font-size: 14px;
  line-height: 20px;
  font-family: OswaldLight,Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: 0;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  color: #999999;
  opacity: 1.0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  position: relative;
  padding: 16px;
  line-height: normal;
  -webkit-font-smoothing: inherit;
  -moz-osx-font-smoothing: inherit;
  -webkit-appearance: none;
}

.c0::after {
  content: '';
  position: absolute;
  display: block;
  bottom: -2px;
  height: 1px;
  left: 0;
  width: calc(100%);
  background-color: #999999;
  -webkit-transform-origin: right center;
  -ms-transform-origin: right center;
  transform-origin: right center;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
  -webkit-transition: -webkit-transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  -webkit-transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
}

.c0:hover::after {
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c0::-moz-focus-inner {
  border: 0;
  padding: 0;
}

.c2 {
  display: block;
  color: #999999;
}

<button
  className="c0"
  color="grey4"
  onClick={[MockFunction]}
>
  <span
    className="c1 c2"
    color="grey4"
    fontSize="xs"
  >
    Button
  </span>
</button>
`;

exports[`<Link /> component should render an anchor tag against the "href" attribute 1`] = `
.c1 {
  color: #999999;
  font-size: 14px;
  line-height: 20px;
  font-family: OswaldLight,Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  color: #999999;
  opacity: 1.0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  position: relative;
  padding: 16px;
}

.c0::after {
  content: '';
  position: absolute;
  display: block;
  bottom: -2px;
  height: 1px;
  left: 0;
  width: calc(100%);
  background-color: #999999;
  -webkit-transform-origin: right center;
  -ms-transform-origin: right center;
  transform-origin: right center;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
  -webkit-transition: -webkit-transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  -webkit-transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
}

.c0:hover::after {
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c2 {
  display: block;
  color: #999999;
}

<a
  className="c0"
  color="grey4"
  href="https://google.com/?how-to-write-tests"
  target="_self"
>
  <span
    className="c1 c2"
    color="grey4"
    fontSize="xs"
  >
    Button
  </span>
</a>
`;

exports[`<Link /> component should render an anchor tag from a router link against the "to" attribute 1`] = `
.c1 {
  color: #999999;
  font-size: 14px;
  line-height: 20px;
  font-family: OswaldLight,Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  color: #999999;
  opacity: 1.0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  position: relative;
  padding: 16px;
}

.c0::after {
  content: '';
  position: absolute;
  display: block;
  bottom: -2px;
  height: 1px;
  left: 0;
  width: calc(100%);
  background-color: #999999;
  -webkit-transform-origin: right center;
  -ms-transform-origin: right center;
  transform-origin: right center;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
  -webkit-transition: -webkit-transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  -webkit-transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
}

.c0:hover::after {
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c2 {
  display: block;
  color: #999999;
}

<a
  className="c0"
  href="/"
  onClick={[Function]}
>
  <span
    className="c1 c2"
    color="grey4"
    fontSize="xs"
  >
    Button
  </span>
</a>
`;

exports[`<Link /> component should render an anchor tag with a left icon 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 svg {
  width: 25px;
  height: 25px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  fill: currentColor;
}

.c2 svg g,
.c2 svg p {
  fill: currentColor;
}

.c3 {
  color: #999999;
  font-size: 16px;
  line-height: 24px;
  font-family: OswaldLight,Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  color: #999999;
  opacity: 1.0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  position: relative;
  padding: 16px;
}

.c0::after {
  content: '';
  position: absolute;
  display: block;
  bottom: -2px;
  height: 1px;
  left: 24px;
  width: calc( 100% - 24px );
  background-color: #999999;
  -webkit-transform-origin: right center;
  -ms-transform-origin: right center;
  transform-origin: right center;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
  -webkit-transition: -webkit-transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  -webkit-transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
}

.c0:hover::after {
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c1 {
  margin-left: 0px;
  margin-right: 4px;
}

.c1 svg {
  height: 20px;
  width: 20px;
}

.c4 {
  display: block;
  color: #999999;
}

<a
  className="c0"
  color="grey4"
  target="_self"
>
  <span
    className="c1"
    color="grey4"
  >
    <div
      className="c2"
    >
      <svg
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.78 11.28a.75.75 0 01.22.53v.38a.77.77 0 01-.22.53l-5.14 5.13a.5.5 0 01-.71 0l-.71-.71a.49.49 0 010-.7L13.67 12 9.22 7.56a.5.5 0 010-.71l.71-.7a.5.5 0 01.71 0z"
        />
      </svg>
    </div>
  </span>
  <span
    className="c3 c4"
    color="grey4"
    fontSize="sm"
  >
    Button
  </span>
</a>
`;

exports[`<Link /> component should render an anchor tag with a right icon 1`] = `
.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 svg {
  width: 25px;
  height: 25px;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  fill: currentColor;
}

.c4 svg g,
.c4 svg p {
  fill: currentColor;
}

.c1 {
  color: #999999;
  font-size: 14px;
  line-height: 20px;
  font-family: OswaldLight,Arial,sans-serif;
  font-weight: 400;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  color: #999999;
  opacity: 1.0;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  position: relative;
  padding: 16px;
}

.c0::after {
  content: '';
  position: absolute;
  display: block;
  bottom: -2px;
  height: 1px;
  width: calc( 100% - 20px );
  background-color: #999999;
  -webkit-transform-origin: right center;
  -ms-transform-origin: right center;
  transform-origin: right center;
  -webkit-transform: scale(0,1);
  -ms-transform: scale(0,1);
  transform: scale(0,1);
  -webkit-transition: -webkit-transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  -webkit-transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
  transition: transform 0.25s cubic-bezier(0.37,0.31,0.2,0.85);
}

.c0:hover::after {
  -webkit-transform-origin: left center;
  -ms-transform-origin: left center;
  transform-origin: left center;
  -webkit-transform: scale(1,1);
  -ms-transform: scale(1,1);
  transform: scale(1,1);
}

.c3 {
  margin-left: 4px;
  margin-right: 0px;
}

.c3 svg {
  height: 16px;
  width: 16px;
}

.c2 {
  display: block;
  color: #999999;
}

<a
  className="c0"
  color="grey4"
  target="_self"
>
  <span
    className="c1 c2"
    color="grey4"
    fontSize="xs"
  >
    Button
  </span>
  <span
    className="c3"
    color="grey4"
  >
    <div
      className="c4"
    >
      <svg
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M15.78 11.28a.75.75 0 01.22.53v.38a.77.77 0 01-.22.53l-5.14 5.13a.5.5 0 01-.71 0l-.71-.71a.49.49 0 010-.7L13.67 12 9.22 7.56a.5.5 0 010-.71l.71-.7a.5.5 0 01.71 0z"
        />
      </svg>
    </div>
  </span>
</a>
`;
